.gitlab-ci.yml

```diff
@@ -1,259 +1,367 @@
+# | PE 2016.4 | 4.7*   | 2.1  | 2018-10
+# | PE 2016.5 | 4.8    | 2.1  | 2017-05
+# | SIMP 6.0  | 4.8    | 2.1  | TBD
+# | PE 2017.1 | 4.9    | 2.1  | 2017-10
+# | PE 2017.2 | 4.10   | 2.1  | 2018-02
+# | SIMP 6.1  | 4.10   | 2.1  | TBD
+# | PE 2017.3 | 5.3    | 2.4  | 2018-08
+# | PE 2018.1 | 5.5    | 2.4  | 2020-05
+#
+# * PE 2016.4 released with Puppet 4.7.0, but upgraded to the 4.10 series
+# starting with 2016.4.5.
+#
+
+.pe_2016_4: &pe_2016_4
   variables:
     PUPPET_VERSION: '~> 4.7.0'
-  <<: *cache_bundler
-  <<: *setup_bundler_env
-  <<: *spec_tests
 
+.simp_6_0: &simp_6_0
+  variables:
+    PUPPET_VERSION: '~> 4.8.1'
 
-# Puppet 4.8 for SIMP 6.0 + 6.1 support
-# --------------------------------------
-pup4_8-validation:
-  stage: validation
-  tags:
-    - docker
-  image: ruby:2.1
+.simp_6_1: &simp_6_1
   variables:
-    PUPPET_VERSION: '~> 4.8.0'
-  <<: *cache_bundler
-  <<: *setup_bundler_env
-  <<: *validation_checks
+    PUPPET_VERSION: '~> 4.10.8'
 
-pup4_8-unit:
-  stage: unit
-  tags:
-    - docker
-  image: ruby:2.1
+.pe_2017_3: &pe_2017_3
   variables:
-    PUPPET_VERSION: '~> 4.8.0'
-  <<: *cache_bundler
-  <<: *setup_bundler_env
-  <<: *spec_tests
+    PUPPET_VERSION: '~> 5.3.2'
 
+.pe_2018_1: &pe_2018_1
+  variables:
+    PUPPET_VERSION: '~> 5.5.1'
 
-# Puppet 4.10 for PE 2017.2 support (EOL:2018-02-21)
-# See: https://puppet.com/misc/puppet-enterprise-lifecycle
-# --------------------------------------
-pup4_10-validation:
-  stage: validation
-  tags:
-    - docker
-  image: ruby:2.1
+.pup4_latest: &pup4_latest
   variables:
-    PUPPET_VERSION: '~> 4.10.0'
-  <<: *cache_bundler
-  <<: *setup_bundler_env
-  <<: *validation_checks
+    PUPPET_VERSION: '~> 4.0'
 
-pup4_10-unit:
-  stage: unit
-  tags:
-    - docker
-  image: ruby:2.1
+.pup5_latest: &pup5_latest
   variables:
-    PUPPET_VERSION: '~> 4.10.0'
-  <<: *cache_bundler
-  <<: *setup_bundler_env
-  <<: *spec_tests
+    PUPPET_VERSION: '~> 5.0'
 
 
-# Puppet 5.3 for PE 2017.3 support (EOL: 2018-07)
-# See: https://puppet.com/misc/puppet-enterprise-lifecycle
-# --------------------------------------
-pup5_3-validation:
-  stage: validation
-  tags:
-    - docker
-  image: ruby:2.4
-  variables:
-    PUPPET_VERSION: '~> 5.3.0'
-  <<: *cache_bundler
-  <<: *setup_bundler_env
-  <<: *validation_checks
-  allow_failure: true
+#=======================================================================
+# Basic Sanity Checks
+#
 
-pup5_3-unit:
-  stage: unit
-  tags:
-    - docker
-  image: ruby:2.4
-  variables:
-    PUPPET_VERSION: '~> 5.3.0'
-  <<: *cache_bundler
-  <<: *setup_bundler_env
-  <<: *spec_tests
-  allow_failure: true
+# Execute simple sanity checks on the environment before proceeding to more
+# resource-intensive jobs.  Besides running checks, this condenses the initial
+# cache generation into a single job for the later stages.  The first stage,
+# in particular, would otherwise suffer a heavy cache-miss penalty as its
+# jobs kick off in parallel.
```

Comment:
I've updated it to use the `CI_PROJECT_PATH_SLUG` so it persists across PRs for the entire repo.  I didn't want to use the `CI_NAMESPACE_*` (group-level) variables, because there may be legitimate reasons to use different cache content from repo to repo.  It might be worth sharing those across repo types, i.e., pupmod and rubygem, but I thought that should/could wait.  Also, the upload and download of the cache file takes a lot of time, so I wanted to prevent it from growing forever and ever.  That, too, could use some work outside of this PR, maybe in how the runners are configured, but for now I felt this was the most expedient way to ensure the cache expired.